Support sorting by last QA started time

ikreymer opened the pull request:
To support #1683, it would be useful to be able to sort by 'last QA start time' in addition to/instead of last QA state.

ikreymer committed:
sort by lastQAStarted

tw4l commented on backend/btrixcloud/basecrawls.py:
```diff
@@ -634,6 +636,10 @@ async def list_all_base_crawls(
                 sort_query["type"] = 1
 
+            if sort_by == "qaStarted":
+                sort_query["lastQAStarted"] = sort_direction
```
I don't think this bit is quite right, though we probably want to keep the secondary `sort_query["type"] = 1` check. As-is, the main sort we want (`lastQAStarted`) will be a secondary sort, and the primary sort will be on a field that doesn't exist (`qaStarted`). Which doesn't take into account the active QA runs yet.

tw4l replied:
We do it like the `qaState` where the active `qaState` is internally keyed to active, then we add secondary sort on last (finished) state. Which is necessary because we have `qa` and `qaFinished` separated in the model.

Or we could add active qas to the qaFinished array we're generating prior to sorting by starting time - maybe that's simpler, since we don't care about whether it's finished or not in this case. Leaning toward implementing that way now that I'm thinking about it more.

tw4l replied:
Also might be simpler if the data returned matching the sort key we're asking for. So maybe we use `lastQAStarted` as the sort key in the query params?

tw4l commented on backend/btrixcloud/crawls.py:
```diff
@@ -251,6 +253,9 @@ async def list_crawls(
                 sort_query["lastQAState"] = sort_direction
 
+            if sort_by == "qaStarted":
+                sort_query["lastQAStarted"] = sort_direction
```
Same issue as above with basecrawls - we want the primary sort to be on `lastQAStarted`

tw4l commented on backend/test/test_qa.py:
```diff
@@ -138,7 +139,7 @@ def failed_qa_run_id(crawler_crawl_id, crawler_auth_headers, default_org_id):
     assert crawls[0]["id"] == crawler_crawl_id
     assert crawls[0]["activeQAState"]
     assert crawls[0]["activeQAStats"]
-    assert crawls[0]["lastQAState"]
```
I think we can keep this in? Also should add another test with the right query params.

tw4l committed:
temp

tw4l committed:
Rename qa sorts to lastQAStarted, lastQAState

Include active QA in these (will already become first since we're
sorted by started) instead of separating it out into a separate
field. This is more consistent with how we handle workflows.

We do however keep activeQAStats, as that can be used to quickly
determine if there is an active QA run and show a progress bar for
it if so.

tw4l commented:
@ikreymer I took a crack at simplifying the QA sorting fields and making them more consistent with what we do for workflows. Namely:

- `sortBy` fields renamed to `lastQAState` and `lastQAStarted`
- Current QA runs are now included in the `lastQAState`/`lastQAStarted` fields, rather than being separated out to different values
- Kept `activeQAStats` as-is for progress bar.

Ready for your eyes.

This will require some small tweaks to #1683 , will push those next. And if we want to revert the changes, can always just drop the commit.

tw4l committed:
Fix aggregation - add missing dollar sign to set

tw4l committed:
Update sortBys in upload sorting test

tw4l commented:
#1683 updated to match!

ikreymer commented:
Thanks for the update! Only question is if new operators like concatArrays would effect perf, but since this is all on one object, hopefully fairly negligible.

tw4l commented:
> Thanks for the update! Only question is if new operators like concatArrays would effect perf, but since this is all on one object, hopefully fairly negligible.

FWIW they definitely do to some degree but we can see if that becomes a problem. Likely the most performant solution longterm is to precompute more of these values, but that comes with its own complexity tradeoff so I'd wait to see if it becomes an issue.
